article-api/src/main/scala/no/ndla/articleapi/db/migration/V45__MigrateMarkdown.scala

```diff
@@ -0,0 +1,151 @@
+    sql"select id, document from contentdata where document is not null order by id limit 1000 offset $offset"
+      .map(rs => {
+        (rs.long("id"), rs.string("document"))
+      })
+      .list()
+  }
+
+  private def updateArticle(document: String, id: Long)(implicit session: DBSession): Int = {
+    val dataObject = new PGobject()
+    dataObject.setType("jsonb")
+    dataObject.setValue(document)
+
+    sql"update contentdata set document = $dataObject where id = $id".update()
+  }
+
+  def updateContent(contents: JArray, contentType: String, function: String => String): json4s.JValue = {
+    contents.map(content =>
+      content.mapField {
+        case (`contentType`, JString(html)) => (`contentType`, JString(function(html)))
+        case z                              => z
+      }
+    )
+  }
+
+  private def stringToJsoupDocument(htmlString: String): Element = {
+    val document = Jsoup.parseBodyFragment(htmlString)
+    document.outputSettings().escapeMode(EscapeMode.xhtml).prettyPrint(false)
+    document.select("body").first()
+  }
+
+  private def jsoupDocumentToString(element: Element): String = {
+    element.select("body").html()
+  }
+
+  private def fixCaption(html: String): String = {
+    val doc = stringToJsoupDocument(html)
+    doc
+      .select("ndlaembed")
+      .forEach(embed => {
+        val hasCaption = embed.hasAttr("data-caption")
+
+        if (hasCaption) {
+          val oldCaption = embed.attr("data-caption")
+          val newCaption = convertMarkdown(oldCaption)
+          embed.attr("data-caption", newCaption): Unit
+        }
+
+      })
+    jsoupDocumentToString(doc)
+  }
+
+  private def stripOuterParagraph(html: String): String = {
+    val doc = stringToJsoupDocument(html).selectFirst("body")
+    if (doc.childrenSize() == 1 && doc.child(0).tagName() == "p") {
+      doc.child(0).unwrap()
+    }
+    jsoupDocumentToString(doc)
+  }
+
+  private[migration] def convertDocument(document: String): String = {
+    val oldArticle = parse(document)
+
+    val newArticle = oldArticle.mapField {
+      case ("content", contents: JArray) =>
+        val updated = updateContent(contents, "content", fixCaption)
+        ("content", updated)
+      case ("introduction", introductions: JArray) =>
+        val updated = updateContent(introductions, "introduction", convertMarkdown)
+        ("introduction", updated)
+      case ("visualElement", visualElements: JArray) =>
+        val updated = updateContent(visualElements, "visualElement", fixCaption)
+        ("visualElement", updated)
+
+      case x => x
+    }
+
+    compact(render(newArticle))
+  }
+
+  private def convertMarkdown(content: String): String = {
+    val options = new MutableDataSet()
+    options.set(HtmlRenderer.SOFT_BREAK, "<br>")
+    // Cast the extensions to Extension before adding them to the list
+    val extensions =
+      List(SubscriptExtension.create(), SuperscriptExtension.create()).map(_.asInstanceOf[Extension]).asJava
```

Comment:
Er vel ikke så fryktelig farlig siden det bare er migreringer, men hadde vært mer scala'ete å gjort sånn:
```suggestion
    val extensions = List[Extension](SubscriptExtension.create(), SuperscriptExtension.create()).asJava
```